### PR TITLE
vNextMobile

### PR DESCRIPTION
**Pull request by kauefelipe**
- Refactored all mobile login (thanks to Hamala that send me all implementation) 
- AuctionResponse and WatchListResponse inherits IListResponse
- Added several missed attributes to Item and ItemData
- Corrected playerSerach to ignore resourceId if not setted
- Updated Newtosoft.Json library to 9.0.1

### CODE REVIEW

**Comment by trydis** [comment]
Appreciated 👍 I have a few questions/comments before merging, hope that's ok.

File: UltimateTeam.Toolkit/Models/IListResponse.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+
+namespace UltimateTeam.Toolkit.Models
+{
+    public interface IListResponse
```
Comment (trydis) [inline UltimateTeam.Toolkit/Models/IListResponse.cs]: What's the purpose of this interface? `AuctionResponse` and `WatchlistResponse` could share a base class or similar, but can't see why we need this.

File: UltimateTeam.Toolkit/Models/AuctionResponse.cs [hunk 2]
```diff
@@ -2,13 +2,13 @@
 
 namespace UltimateTeam.Toolkit.Models
 {
-    public class AuctionResponse
+    public class AuctionResponse : IListResponse
```
Comment (trydis) [inline UltimateTeam.Toolkit/Models/AuctionResponse.cs]: Why are 2 properties `virtual` here? Why should it be possible to override them?

File: UltimateTeam.Toolkit/Models/WatchlistResponse.cs [hunk 3]
```diff
@@ -2,13 +2,13 @@
 
 namespace UltimateTeam.Toolkit.Models
 {
-    public class WatchlistResponse
+    public class WatchlistResponse : IListResponse
```
Comment (trydis) [inline UltimateTeam.Toolkit/Models/WatchlistResponse.cs]: Why are 3 properties `virtual` here (only 2 were `virtual` in `AuctionResponse`)? Why should it be possible to override them?

File: UltimateTeam.Toolkit/Models/IListResponse.cs (hunk 1 above)
Comment (kauefelipe) [inline UltimateTeam.Toolkit/Models/IListResponse.cs]: Hi Trydis...
In my BOT I buy several players, some by normal bid others by buy now.
After the round of these bids I call a method (SendDuplicatedItemToClub) that receives this interface and sends all the duplicate players in the list to my club (I keep it just in case one day needs for some SBC for example). 

I created the interface only to need to implement the method only once since both of these types have this list of "duplicates". It really is not mandatory in case you want to retire feel free.

File: UltimateTeam.Toolkit/Models/AuctionResponse.cs (hunk 2 above)
Comment (kauefelipe) [inline UltimateTeam.Toolkit/Models/AuctionResponse.cs]: I confess I do not remember why I did this. Lol
It must have been some crazy thing that I started to do, but I gave up. It really does not make much sense for the methods to be virtual in this case.

Comment (kauefelipe) [inline UltimateTeam.Toolkit/Models/AuctionResponse.cs]: I confess I do not remember why I did this. Lol
It must have been some crazy thing that I started to do, but I gave up. It really does not make much sense for the methods to be virtual in this case.

File: UltimateTeam.Toolkit/Models/WatchlistResponse.cs (hunk 3 above)
Comment (kauefelipe) [inline UltimateTeam.Toolkit/Models/WatchlistResponse.cs]: Again I do not know. Can you ignore or prefer that I send a new PR?

**Comment by kauefelipe** [comment]
Just to give credits and thank, the mobile part was implemented by the @Lululuz ! :smile:

File: UltimateTeam.Toolkit/Models/IListResponse.cs (hunk 1 above)
Comment (trydis) [inline UltimateTeam.Toolkit/Models/IListResponse.cs]: I see, thanks for explaining. I would prefer to keep the toolkit generic and not introduce too many specific interfaces etc. What i could do is create a base class for `AuctionResponse` and `WatchlistResponse` which has the same properties as the interface. Then you can use that in your method instead. Almost the same, but reduces some duplicated code and less confusing i think.

Would be great if you could push another commit to your branch that removes the interface.

File: UltimateTeam.Toolkit/Models/AuctionResponse.cs (hunk 2 above)
Comment (trydis) [inline UltimateTeam.Toolkit/Models/AuctionResponse.cs]: Would be great if you could push another commit to your branch that removes ´virtual`.

File: UltimateTeam.Toolkit/Models/WatchlistResponse.cs (hunk 3 above)
Comment (trydis) [inline UltimateTeam.Toolkit/Models/WatchlistResponse.cs]: Would be great if you could push another commit to your branch that removes ´virtual`. Just push to the same branch and they will show up here.

I can change it locally if you're not able to push another commit, your choice :smile:

**Comment by trydis** [comment]
Will credit you both 👍

**Comment by trydis** [comment]
Nuget prerelease package, let me know how it works: https://www.nuget.org/packages/UltimateTeam.Toolkit/5.0.0-rc0001

**Comment by A-VORONKIN** [comment]
It not handle errors from mobile errors. Code 461 didn't throw PermissionDenied Exception.

**Comment by Lululuz** [comment]
The problem is that mobile errors are thrown as actual response codes. Webapp always returns code 200 also when you get a permission denied error. Mobile set the status code 461 and therefore the `message.EnsureSuccessStatusCode();` throws an unspecified exception.

I solved it like this. Not nice but it works. 

```
private static void CheckForMobileException(HttpResponseMessage message)
        {
            switch ((ushort)message.StatusCode)
            {
                case 200:
                    break;
                case (ushort)FutErrorCode.ExpiredSession:
                    throw new ExpiredSessionException(new FutError { Code = FutErrorCode.ExpiredSession }, new Exception(FutErrorCode.ExpiredSession.ToString()));
                case (ushort)FutErrorCode.NotFound:

                    ...

                default:
                    throw new UnkownHttpResponseCodeException((ushort) message.StatusCode);
            }
        }

protected static void EnsureSuccessStatusCode(HttpResponseMessage message)
        {
            CheckForMobileException(message);
            message.EnsureSuccessStatusCode();
        }
```

I then replaced every `message.EnsureSuccessStatusCode();` with my custom one. For example:

```
var removeFromTradePileMessage = await removeFromTradePileMessageTask.ConfigureAwait(false);
EnsureSuccessStatusCode(removeFromTradePileMessage);
```

But maybe someone else has a better idea and more patience :)

@kauefelipe Sorry forgot to send you this

**Comment by trydis** [comment]
I haven't tried it, but would something like this work?

```csharp
protected static async Task<T> DeserializeAsync<T>(HttpResponseMessage message) where T : class
{
    string messageContent = null;

    if (message.IsSuccessStatusCode)
    {
        messageContent = await message.Content.ReadAsStringAsync();
    }
    else
    {
        MapAndThrowException(null, new FutError {Code = (FutErrorCode)message.StatusCode });
    }

    T deserializedObject = null;

    try
    {
        deserializedObject = JsonConvert.DeserializeObject<T>(messageContent, JsonSerializerSettings);
    }
    catch (JsonSerializationException serializationException)
    {
        try
        {
            var futError = JsonConvert.DeserializeObject<FutError>(messageContent, JsonSerializerSettings);
            MapAndThrowException(serializationException, futError);
        }
        catch (JsonSerializationException)
        {
            throw serializationException;
        }
    }

    return deserializedObject;
}
```

**Comment by Lululuz** [comment]
Yes, looks like a much easier and nicer solution than mine 👍. You'd still need to remove the `xxMessage.EnsureSuccessStatusCode();` from every request that uses it thought.

**Comment by trydis** [comment]
Good point @Lululuz

**Comment by alraik** [comment]
hi can give me your skype need big site

**Comment by tringler** [comment]
@kauefelipe Are you interested to adjust your PR to the master branch as soon we pushed the FUT18 adjustments to the master branch?
